src/V3RandomizeMethod.cpp

```diff
@@ -0,0 +1,199 @@
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3RandomizeMethod.h"
+
+//######################################################################
+// Visitor that marks classes needing a randomize() method
+
+class RandomizeMethodMarkVisitor : public AstNVisitor {
+private:
+    // NODE STATE
+    // Cleared on Netlist
+    //  AstClass::user1()       -> bool.  Set true to indicate needs randomize processing
+    AstUser1InUse m_inuser1;
+
+    typedef std::unordered_set<AstClass*> DerivedSet;
+    typedef std::unordered_map<AstClass*, DerivedSet> BaseToDerivedMap;
+
+    BaseToDerivedMap m_baseToDerivedMap;  // Mapping from base classes to classes that extend them
+
+    // METHODS
+    VL_DEBUG_FUNC;
+
+    void markMembers(AstClass* nodep) {
+        for (auto* memberp = nodep->stmtsp(); memberp; memberp = memberp->nextp()) {
+            // If member is rand and of class type, mark its class
+            if (VN_IS(memberp, Var) && VN_CAST(memberp, Var)->isRand()) {
+                if (auto* classRefp = VN_CAST(memberp->dtypep(), ClassRefDType)) {
+                    auto* classp = classRefp->classp();
+                    markMembers(classp);
+                    markDerived(classp);
+                    classRefp->classp()->user1(true);
+                }
+            }
+        }
+    }
+
+    void markDerived(AstClass* nodep) {
+        if (m_baseToDerivedMap.find(nodep) != m_baseToDerivedMap.end()) {
+            for (auto* classp : m_baseToDerivedMap.at(nodep)) {
+                classp->user1(true);
+                markMembers(classp);
+                markDerived(classp);
+            }
+        }
+    }
+
+    void markAllDerived() {
+        for (auto p : m_baseToDerivedMap) {
+            if (p.first->user1()) { markDerived(p.first); }
+        }
+    }
+
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        iterateChildren(nodep);
+        if (nodep->extendsp()) {
+            // Save pointer to derived class
+            auto* basep = nodep->extendsp()->classp();
+            m_baseToDerivedMap[basep].insert(nodep);
+        }
+    }
+
+    virtual void visit(AstMethodCall* nodep) override {
+        iterateChildren(nodep);
+        if (nodep->name() != "randomize") return;
+        if (AstClassRefDType* classRefp = VN_CAST(nodep->fromp()->dtypep(), ClassRefDType)) {
+            auto* classp = classRefp->classp();
+            classp->user1(true);
+            markMembers(classp);
+        }
+    }
+
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+public:
+    // CONSTRUCTORS
+    explicit RandomizeMethodMarkVisitor(AstNetlist* nodep) {
+        iterate(nodep);
+        markAllDerived();
+    }
+    virtual ~RandomizeMethodMarkVisitor() override {}
```

Comment:
```suggestion
    virtual ~RandomizeMethodMarkVisitor() override = default;
```

(Recent master change for performance)